File: backend/app/schemas/orbit.py
"""Response models for the casing orbit / Lissajous plot."""
from datetime import datetime
from typing import List, Optional
from uuid import UUID

from pydantic import BaseModel, Field


class CasingOrbitOut(BaseModel):
    upload_id: UUID
    sensor_id: UUID
    captured_at: datetime
    original_filename: Optional[str] = None

    x_channel: int
    y_channel: int
    #: Engineering unit of x_data/y_data. Displacement derived from acceleration.
    amplitude_unit: str = "um"
    source_unit: str = "g"
    #: True when the plotted values are double-integrated displacement.
    is_displacement: bool = True

    sampling_rate_hz: float
    sample_count: int
    record_duration_s: float

    harmonic: int
    centre_hz: float
    lower_hz: float
    upper_hz: float
    frequency_resolution_hz: float
    requested_half_bandwidth_hz: float
    effective_half_bandwidth_hz: float
    effective_bandwidth_percent: float
    bandwidth_sufficient: bool
    required_duration_s: float
    integration_floor_hz: float

    filter_revolutions: int
    available_revolutions: float
    display_revolutions: float
    filter_duration_s: float
    display_duration_s: float

    #: FFT-derived estimate — never a measured tachometer reading.
    estimated_shaft_hz: Optional[float] = None
    shaft_source: str = "estimated_fft"

    x_data: List[float]
    y_data: List[float]
    elapsed_s: List[float]
    revolution: List[float]
    unfiltered_x: Optional[List[float]] = None
    unfiltered_y: Optional[List[float]] = None
    point_count: int
    peak_displacement_um: float

    #: "time_only" until a real keyphasor exists.
    phase_reference: str
    keyphasor_available: bool = False
    #: Numeric probe mounting angles are not stored anywhere, so no X/Y rotation is
    #: applied and the UI must not claim an exact physical orientation.
    mounting_angle_configured: bool = False

    sensor_label: Optional[str] = None
    sensor_orientation: Optional[str] = None
    mounting_location: Optional[str] = None

    warnings: List[str] = Field(default_factory=list)

File: frontend/src/types/orbit.ts
/** Shapes returned by GET /api/v1/measurements/uploads/{id}/orbit. */

export const ORBIT_HARMONICS = [1, 2] as const;
export type OrbitHarmonic = (typeof ORBIT_HARMONICS)[number];

export const ORBIT_DISPLAY_REVS = [1, 2, 4, 8] as const;
export const ORBIT_FILTER_REVS = [4, 8, 16, 32] as const;
export const ORBIT_BANDWIDTHS = [0.5, 2, 5, 10, 20] as const;

export interface CasingOrbitResponse {
  upload_id: string;
  sensor_id: string;
  captured_at: string;
  original_filename: string | null;

  x_channel: number;
  y_channel: number;
  amplitude_unit: string;
  source_unit: string;
  is_displacement: boolean;

  sampling_rate_hz: number;
  sample_count: number;
  record_duration_s: number;

  harmonic: number;
  centre_hz: number;
  lower_hz: number;
  upper_hz: number;
  frequency_resolution_hz: number;
  requested_half_bandwidth_hz: number;
  effective_half_bandwidth_hz: number;
  effective_bandwidth_percent: number;
  bandwidth_sufficient: boolean;
  required_duration_s: number;
  integration_floor_hz: number;

  filter_revolutions: number;
  available_revolutions: number;
  display_revolutions: number;
  filter_duration_s: number;
  display_duration_s: number;

  /** FFT-derived estimate — never a measured tachometer reading. */
  estimated_shaft_hz: number | null;
  shaft_source: string;

  x_data: number[];
  y_data: number[];
  elapsed_s: number[];
  revolution: number[];
  unfiltered_x: number[] | null;
  unfiltered_y: number[] | null;
  point_count: number;
  peak_displacement_um: number;

  phase_reference: string;
  keyphasor_available: boolean;
  mounting_angle_configured: boolean;

  sensor_label: string | null;
  sensor_orientation: string | null;
  mounting_location: string | null;

  warnings: string[];
}

export interface CasingOrbitQuery {
  uploadId: string;
  xChannel: number;
  yChannel: number;
  harmonic: number;
  bandwidthPercent: number;
  /** Omit for auto (uses as many revolutions as the capture holds, up to 16). */
  filterRevolutions?: number;
  displayRevolutions: number;
  includeUnfiltered: boolean;
}
